### PR TITLE
[WIP] enhance(BDD): adds installer package for BDD

### PR DESCRIPTION
**Pull request by ashishranjan738**
Signed-off-by: Ashish Ranjan <ashishranjan738@gmail.com>

**What this PR does / why we need it**:
This commit adds installer package for BDD

**Special notes for your reviewer**:

**Checklist:**
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: integration-tests/artifacts/installer/v1alpha1/intaller.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
+package v1alpha1
```
Comment (AmitKumarDas) [inline integration-tests/artifacts/installer/v1alpha1/intaller.go]: Please check the name of this .go file. Spellcheck!

File: integration-tests/artifacts/installer/v1alpha1/intaller.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+	if err != nil {
+		return err
+	}
+	err = k.Create(i.object)
+	return err
+}
+
+// Delete triggers delete process for the
+// install client
+func (i *installer) Delete() error {
```
Comment (AmitKumarDas) [inline integration-tests/artifacts/installer/v1alpha1/intaller.go]: can we rename this to `UnInstall`

File: integration-tests/artifacts/installer/v1alpha1/intaller.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+		if b != nil {
+			b(i)
+		}
+	}
+	return i
+}
+
+// FromObject instructs the installer to install
+// the object from a unstructured instance
+func (i *installer) FromObject(u *unstructured.Unstructured) *installer {
```
Comment (AmitKumarDas) [inline integration-tests/artifacts/installer/v1alpha1/intaller.go]: can we have below:
```go
type installer struct {
  object *unstructured.Unstructured
}

type predicate func(*installer) bool

func (i *installer) String() string {
  if i.object == nil {
    return "installer - object=nil"
  }
  return fmt.Sprintf("installer - name=%s namespace=%s kind=%s",
    i.object.GetName(),
    i.object.GetNamespace(),
    i.object.GetKind()
}

func (i *installer) GoString() string {
  return i.String()
}

func (i *installer) Install() error {
  // ...
}

func (i *installer) UnInstall() error {
  // ...
}

func (i *installer) Verify() error {
  // ...
}

type builder struct {
  installer *installer
  errs       []error
}

func BuilderForObject(obj *unstructured.Unstructured) *builder {
  b := &builder{}
  if obj == nil {
    b.errs = append(b.errs, errors.Errorf("failed to build for object: nil unstruct instance provided"))
    return b
  }
  b.installer = &installer{object: obj}}
  return b
}

func BuilderForYaml(yaml string) *builder {
  b := &builder{}
  obj, err := unstruct.builderForYaml(yaml).Build()
  if err != nil {
    b.errors = append(b.errors, err)
    return b
  }
  b.installer = &installer{object: obj}
  return b
}

func IsPodRunning() predicate {
  // ...
}

func IsPodRunningIfLabel(label string) predicate {
  // ...
}
```

File: integration-tests/artifacts/installer/v1alpha1/intaller.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+type listInstaller struct {
+	errors     []error
+	installers []*installer
+	clientset  client.Client
+	namespace  string
+}
+
+// ListInstaller returns a new instance of
+// listInstaller
+func ListInstaller(lbos ...listBuildOptions) *listInstaller {
```
Comment (AmitKumarDas) [inline integration-tests/artifacts/installer/v1alpha1/intaller.go]: Redesign this based on earlier comments

File: integration-tests/artifacts/installer/v1alpha1/intaller.go (hunk 1 above)
Comment (ashishranjan738) [reply integration-tests/artifacts/installer/v1alpha1/intaller.go]: done

File: integration-tests/artifacts/installer/v1alpha1/intaller.go (hunk 2 above)
Comment (ashishranjan738) [reply integration-tests/artifacts/installer/v1alpha1/intaller.go]: done

File: integration-tests/artifacts/installer/v1alpha1/intaller.go (hunk 3 above)
Comment (ashishranjan738) [reply integration-tests/artifacts/installer/v1alpha1/intaller.go]: done

File: integration-tests/artifacts/installer/v1alpha1/intaller.go (hunk 4 above)
Comment (ashishranjan738) [reply integration-tests/artifacts/installer/v1alpha1/intaller.go]: done

File: integration-tests/artifacts/installer/v1alpha1/intaller.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+		b.errs = append(b.errs, errors.Errorf("failed to build for object: nil unstruct instance provided"))
+		return b
+	}
+	b.installer = &installer{object: obj}
+	return b
+}
+
+// BuilderForYaml creates the installer builder from the
+// YAML provided
+func BuilderForYaml(yaml string) *builder {
```
[the remark this answers is not in the document]
Comment (ashishranjan738) [reply integration-tests/artifacts/installer/v1alpha1/intaller.go]: done

File: integration-tests/artifacts/installer/v1alpha1/intaller.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+func (b *builder) Build() (*installer, error) {
+	if len(b.errs) > 0 {
+		return nil, errors.Errorf("%v", b.errs)
+	}
+	return b.installer, nil
+}
+
+// IsPodRunning returns true if the pod is in running
+// state
+func IsPodRunning() predicate {
```
[the remark this answers is not in the document]
Comment (ashishranjan738) [reply integration-tests/artifacts/installer/v1alpha1/intaller.go]: done

File: integration-tests/artifacts/installer/v1alpha1/intaller.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+// builder abstracts the construction of the builder
+type builder struct {
+	installer *installer
+	errs      []error
+}
+
+// BuilderForObject creates the installer builder from
+// the object provided
+func BuilderForObject(obj *unstructured.Unstructured) *builder {
```
[the remark this answers is not in the document]
Comment (ashishranjan738) [reply integration-tests/artifacts/installer/v1alpha1/intaller.go]: done

File: pkg/unstruct/v1alpha2/kubernetes.go [hunk 8]
```diff
@@ -17,43 +17,87 @@ limitations under the License.
...
+			return cli.Resource(clientset.GroupVersionResourceFromGVK(obj)).Namespace(obj.GetNamespace()).Delete(obj.GetName(), opts, subresources...)
+		}
 	}
 	return nil
 }
 
 // WithClient sets the kubernetes client against
 // the kubeclient instance
-func WithClient(c client.Client) kubeclientBuildOption {
+func WithClient(c k8sdynamic.Interface) kubeclientBuildOption {
```
[the remark this answers is not in the document]
Comment (ashishranjan738) [reply pkg/unstruct/v1alpha2/kubernetes.go]: done

File: pkg/unstruct/v1alpha1/unstruct.go [hunk 9]
```diff
@@ -17,67 +17,122 @@ limitations under the License.
...
+		}
+		lb.list.items = append(lb.list.items, a)
+	}
+	return lb
+}
+
+// ListBuilderForObjects returns a mew instance of
+// list unstruct builder by making use of the provided
+// unstructured object
+func ListBuilderForObjects(objs ...*unstructured.Unstructured) *listBuilder {
```
[the remark this answers is not in the document]
Comment (ashishranjan738) [reply pkg/unstruct/v1alpha1/unstruct.go]: done

File: pkg/unstruct/v1alpha1/unstruct.go [hunk 10]
```diff
@@ -17,67 +17,122 @@ limitations under the License.
...
+// listBuilder enables building a list
+// of an unstruct instance
+type listBuilder struct {
+	list unstructList
+	errs []error
+}
+
+// ListBuilderForYamls returns a mew instance of
+// list unstruct builder by making use of the provided YAMLs
+func ListBuilderForYamls(docs string) *listBuilder {
```
[the remark this answers is not in the document]
Comment (ashishranjan738) [reply pkg/unstruct/v1alpha1/unstruct.go]: done

**Pull request closed without merging**